### DISCUSSION TITLE
Why stair-like loss curve?

### DISCUSSION THREAD

**Discussion by ChenDelong1999** [mlfoundations/open_clip#262] [Q&A]
As in [here](https://github.com/mlfoundations/open_clip/blob/main/docs/clip_conceptual_captions.md#training-loss-curve) as well as my own implementation, stair-like loss curves are observed. Any possible reason for this?

**Comment by rwightman** [comment]
I can't speak to the training run for the graph as I didn't do them, @mitchellnw would have a better idea... but looks like it could be a shuffling issue (as in not properly shuffling)

**Comment by mitchellnw** [comment]
My guess is also a shuffling issue with webdataset when these were run

**Comment by rom1504** [comment]
If the data is not preshuffled you need both shards shuffling and local shuffling

**Comment by ChenDelong1999** [comment]
Perhaps it is not a shuffling issue with webdataset, since I train the model on CC3M (csv dataset), and observed the following curves.

<img width="577" alt="image" src="https://user-images.githubusercontent.com/31244423/171557357-77c1f1a8-1fc8-4ac3-9f2a-d4d0bd2fabc2.png">

which looks very similar to [this curve](https://github.com/mlfoundations/open_clip/blob/main/docs/clip_conceptual_captions.md#training-loss-curve) in `open_clip/docs/clip_conceptual_captions.md`:
![image](https://user-images.githubusercontent.com/31244423/171557622-26119e96-b21d-4460-a189-1ced78f2a202.png)

Loss increases within each epoch, then decrease after each epoch...

**Comment by rom1504** [comment]
@ChenDelong1999 did you preshuffle the dataset (sort randomly the dataset) ?

**Comment by rwightman** [comment]
CsvDataset should be shuffled every epoch, pre-shuffle isn' really relevant. Might be worth checking that https://github.com/mlfoundations/open_clip/blob/d9ee4aa0431b9ea5c99c8617e9b3e0f3f12c458f/src/training/train.py#L62 set_epoch is def being called in distributed case...

**Comment by mitchellnw** [comment]
Looking at this again I wonder if it is caused by the `scale` param, which also exhibits stair-like behaviour.

I would expect that stair-like `scale` => stair-like loss.

But I have no guesses for why `scale` has stair-like behaviour.

To test this hypothesis I would use a 10x smaller learning rate on the `scale` parameter and see if this resolves the issue.

**Comment by rwightman** [comment]
@mitchellnw I've noticed that the scale param has interesting relationship with the LR/loss, I wonder if it's almost behaving in a slightly oscillatory control systems fashion. The scale is strongly impacted by the LR as well, if the LR is high enough the scale will not converge to 100 until it lowers

**Comment by mitchellnw** [comment]
Interesting. I wonder how accuracy/loss would be impacted if this learnable param was replaced by a scheduled param---something like 100 - k*cosine_decay(iteration).

**Comment by viyjy** [comment]
> here

Hi, thanks for your answer, but the `scale` param does not exhibit stair-like behaviour during the training process, isn't it? On the other hand, `scale` is learnable, it shouldn't be stair-like, right?

<img width="529" alt="image" src="https://user-images.githubusercontent.com/20442927/187096350-102636e3-d07d-4721-a5ed-a7ed6dee3378.png">

**Comment by JiachengCheng96** [comment]
Hi @mitchellnw @rom1504, I observed a similar stair-like loss curve when training a ViT-B-32 model on the LAION400M. Do you think this is normal? Or could you share one loss curve of any of your models on LAION400M? Any information would be greatly appreciated.

**Reply by mitchellnw** [reply]
are you using `--dataset-resampled`? I think our working hypothesis is that this is a shuffling issue..

**Reply by JiachengCheng96** [reply]
@mitchellnw Thanks for the reply! No, I didn't use `--dataset-resampled`. Do you think such stair-like loss should be expected when training without `--dataset-resampled`?  I am trying to reproduce the VIT-B-32 result on LAION400M using the exact same setting provided in #191 (which didn't use `--dataset-resampled`), but gets accuracy a bit lower than your 62.9%. Hence, I am wondering if there is something wrong with my stair-like loss.
